File: test_icmp.c
// test_icmp.c - Complete TUN ICMP Echo Reply Test
// Build: gcc -o test_icmp test_icmp.c
// Run: sudo ./test_icmp

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/sys_domain.h>
#include <sys/kern_control.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <net/if_types.h>
#include <net/if_dl.h>

#define BUF_SIZE 4096

// UTUN options (from xnu/private/api/if_utun.h)
#ifndef UTUN_OPT_IFNAME
#define UTUN_OPT_IFNAME 2
#endif

// IPv4 header (avoid macro conflicts)
typedef struct {
    uint8_t  vhl;     // version + header length
    uint8_t  tos;     // type of service
    uint16_t len;     // total length
    uint16_t id;      // identification
    uint16_t off;     // fragment offset
    uint8_t  ttl;     // time to live
    uint8_t  proto;   // protocol
    uint16_t sum;     // checksum
    uint32_t src;     // source address
    uint32_t dst;     // destination address
} iphdr_t;

// ICMP header (avoid macro conflicts)
typedef struct {
    uint8_t  type;
    uint8_t  code;
    uint16_t sum;
    uint16_t id;
    uint16_t seq;
} icmphdr_t;

// Create utun socket and get actual interface name
int create_utun_socket(char *ifname, size_t max_len) {
    struct sockaddr_ctl addr;
    int sock;

    // Create control socket
    sock = socket(PF_SYSTEM, SOCK_DGRAM, SYSPROTO_CONTROL);
    if (sock < 0) {
        return -1;
    }

    // Get control info
    struct ctl_info info;
    memset(&info, 0, sizeof(info));
    strlcpy(info.ctl_name, "com.apple.net.utun_control", sizeof(info.ctl_name));

    if (ioctl(sock, CTLIOCGINFO, &info) < 0) {
        close(sock);
        return -1;
    }

    // Connect to utun
    memset(&addr, 0, sizeof(addr));
    addr.sc_len = sizeof(addr);
    addr.sc_family = AF_SYSTEM;
    addr.ss_sysaddr = AF_SYS_CONTROL;
    addr.sc_id = info.ctl_id;
    addr.sc_unit = 0;  // Auto-assign

    if (connect(sock, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        close(sock);
        return -1;
    }

    // Get kernel-assigned interface name
    char assigned_name[IFNAMSIZ];
    socklen_t name_len = sizeof(assigned_name);
    if (getsockopt(sock, SYSPROTO_CONTROL, UTUN_OPT_IFNAME,
                   assigned_name, &name_len) < 0) {
        close(sock);
        return -1;
    }

    strlcpy(ifname, assigned_name, max_len);
    printf("Created utun socket: %s\n", ifname);
    return sock;
}

// Configure interface IP
int configure_ip(const char *ifname, const char *ip) {
    int sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) return -1;

    struct ifreq ifr;
    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ-1);

    struct sockaddr_in *addr = (struct sockaddr_in *)&ifr.ifr_addr;
    addr->sin_family = AF_INET;
    addr->sin_len = sizeof(struct sockaddr_in);
    inet_pton(AF_INET, ip, &addr->sin_addr);

    int ret = ioctl(sock, SIOCSIFADDR, &ifr);
    close(sock);

    if (ret < 0) {
        fprintf(stderr, "SIOCSIFADDR failed: %s\n", strerror(errno));
        return -1;
    }

    printf("Set IP: %s\n", ip);
    return 0;
}

// Configure peer (destination) address
int configure_peer(const char *ifname, const char *peer) {
    int sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) return -1;

    struct ifreq ifr;
    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ-1);

    struct sockaddr_in *addr = (struct sockaddr_in *)&ifr.ifr_dstaddr;
    addr->sin_family = AF_INET;
    addr->sin_len = sizeof(struct sockaddr_in);
    inet_pton(AF_INET, peer, &addr->sin_addr);

    int ret = ioctl(sock, SIOCSIFDSTADDR, &ifr);
    close(sock);

    if (ret < 0) {
        fprintf(stderr, "SIOCSIFDSTADDR failed: %s\n", strerror(errno));
        return -1;
    }

    printf("Set peer: %s\n", peer);
    return 0;
}

// Bring interface up
int interface_up(const char *ifname) {
    int sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) return -1;

    struct ifreq ifr;
    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ-1);

    if (ioctl(sock, SIOCGIFFLAGS, &ifr) < 0) {
        close(sock);
        return -1;
    }

    ifr.ifr_flags |= IFF_UP | IFF_RUNNING;

    if (ioctl(sock, SIOCSIFFLAGS, &ifr) < 0) {
        close(sock);
        return -1;
    }

    close(sock);
    printf("Interface up\n");
    return 0;
}

// Calculate checksum
uint16_t calc_sum(uint16_t *addr, int len) {
    int nleft = len;
    uint32_t sum = 0;
    uint16_t *w = addr;

    while (nleft > 1) {
        sum += *w++;
        nleft -= 2;
    }
    if (nleft == 1) {
        sum += *(uint8_t *)w;
    }
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    return (uint16_t)(~sum);
}

// Convert IP to string
const char *ip2str(uint32_t ip) {
    static char buf[16];
    uint8_t *p = (uint8_t *)&ip;
    snprintf(buf, sizeof(buf), "%d.%d.%d.%d", p[0], p[1], p[2], p[3]);
    return buf;
}

int main() {
    int tun_fd;
    char tun_name[64] = {0};
    unsigned char buf[BUF_SIZE];
    int n;

    printf("=== TUN ICMP Echo Reply Test ===\n\n");

    // Clean up any existing routes to 10.0.0.2
    printf("Cleaning up existing routes...\n");
    system("route -q -n delete -inet 10.0.0.2 2>/dev/null");

    // Create utun socket
    printf("Creating utun socket...\n");
    tun_fd = create_utun_socket(tun_name, sizeof(tun_name));
    if (tun_fd < 0) {
        fprintf(stderr, "Failed to create utun: %s\n", strerror(errno));
        return 1;
    }

    printf("Interface: %s\n", tun_name);

    // Configure IP and peer
    configure_ip(tun_name, "10.0.0.1");
    configure_peer(tun_name, "10.0.0.2");
    interface_up(tun_name);

    // Add route via shell
    char cmd[256];
    snprintf(cmd, sizeof(cmd), "route -q -n add -inet 10.0.0.2/32 -iface %s 2>&1", tun_name);
    printf("Adding route: %s\n", cmd);
    system(cmd);

    // Verify route
    printf("Verifying route:\n");
    system("route -n get 10.0.0.2 2>&1");

    // Set non-blocking
    int flags = fcntl(tun_fd, F_GETFL, 0);
    fcntl(tun_fd, F_SETFL, flags | O_NONBLOCK);

    printf("\nListening for ICMP...\n");
    printf("(Press Ctrl+C to stop)\n\n");

    while (1) {
        n = read(tun_fd, buf, BUF_SIZE);
        if (n < 0) {
            if (errno == EAGAIN) {
                usleep(10000);
                continue;
            }
            fprintf(stderr, "Read error: %s\n", strerror(errno));
            break;
        }

        printf("=== Received %d bytes ===\n", n);

        // Skip macOS utun 4-byte header if present
        int offset = 0;
        if (n >= 4 && buf[0] == 0 && buf[1] == 0) {
            offset = 4;
            printf("Skipped 4-byte utun header\n");
        }

        if (n - offset < 20) {
            printf("Packet too small\n\n");
            continue;
        }

        iphdr_t *ip = (iphdr_t *)(buf + offset);
        int ip_hlen = (ip->vhl & 0x0F) * 4;
        int ip_len = ntohs(ip->len);

        uint32_t src = ip->src;
        uint32_t dst = ip->dst;

        printf("IP: %s -> %s\n", ip2str(src), ip2str(dst));
        printf("Proto: %d (ICMP=1)\n", ip->proto);

        if (ip->proto != IPPROTO_ICMP) {
            printf("Not ICMP, skipping\n\n");
            continue;
        }

        icmphdr_t *icmp = (icmphdr_t *)(buf + offset + ip_hlen);

        printf("ICMP Type: %d (8=echo, 0=reply)\n", icmp->type);

        if (icmp->type != ICMP_ECHO) {
            printf("Not echo request, skipping\n\n");
            continue;
        }

        printf("Echo Request! ID=0x%04X Seq=%d\n",
               ntohs(icmp->id), ntohs(icmp->seq));

        // Build reply: swap IPs, change type to 0
        // CRITICAL: src = original dst, dst = original src
        ip->src = dst;  // Reply source = original destination
        ip->dst = src;  // Reply destination = original source
        icmp->type = ICMP_ECHOREPLY;

        // Recalculate checksums
        ip->sum = 0;
        ip->sum = calc_sum((uint16_t *)ip, ip_hlen);

        int icmp_len = ip_len - ip_hlen;
        icmp->sum = 0;
        icmp->sum = calc_sum((uint16_t *)icmp, icmp_len);

        printf("Reply: %s -> %s\n\n", ip2str(dst), ip2str(src));

        n = write(tun_fd, buf, n);
        if (n < 0) {
            fprintf(stderr, "Write error: %s\n", strerror(errno));
        } else {
            printf("Sent %d bytes\n\n", n);
        }
    }

    close(tun_fd);
    return 0;
}

File: test_icmp.zig
// test_icmp.zig - TUN ICMP Echo Reply Test (100% Pure Zig)
// Build: zig build-exe test_icmp.zig macos_types.zig -lc -I.
// Run: sudo ./test_icmp
//
// This version uses 100% pure Zig for all logic, including macOS-specific
// types and structures for UTUN operations.

const std = @import("std");
const c = @import("std").c;
const macos = @import("macos_types.zig");

const BUF_SIZE = 4096;

// ============================================================================
// Minimal C Declarations (still required for getsockopt)
// ============================================================================

// getsockopt for utun interface name (Zig 0.13.0 doesn't expose this in std.posix)
extern "c" fn getsockopt(
    sock: c_int,
    level: c_int,
    optname: c_int,
    optval: ?*anyopaque,
    optlen: *c_uint,
) c_int;

// C read/write for file descriptors (required for C-created fd compatibility)
extern "c" fn read(fd: c_int, buf: *anyopaque, nbytes: usize) c_int;
extern "c" fn write(fd: c_int, buf: *const anyopaque, nbytes: usize) c_int;
extern "c" var errno: c_int;

// ============================================================================
// Constants
// ============================================================================

const UTUN_OPT_IFNAME = macos.UTUN_OPT_IFNAME;
const SYSPROTO_CONTROL = macos.SYSPROTO_CONTROL;

// ============================================================================
// UTUN Control Info (Pure Zig using c.ioctl)
// ============================================================================

fn ioctl_get_ctl_info(sock: c_int, ctl_name: [*:0]const u8, ctl_id: *u32) c_int {
    var info: macos.ctl_info = .{};
    info.setName(ctl_name);

    const req = @as(c_int, @bitCast(macos.CTLIOCGINFO));
    const ret = c.ioctl(sock, req, &info);
    if (ret < 0) return -1;

    ctl_id.* = info.ctl_id;
    return 0;
}

// ============================================================================
// Get Interface Name via getsockopt (Pure Zig)
// ============================================================================

fn getsockopt_ifname(sock: c_int, ifname: [*]u8, max_len: usize) c_int {
    var name_buf: [64]u8 = undefined;
    var name_len: c_uint = 64;

    const ret = getsockopt(sock, SYSPROTO_CONTROL, UTUN_OPT_IFNAME, &name_buf, &name_len);
    if (ret < 0) return -1;

    const copy_len = @min(@as(usize, @intCast(name_len)), max_len - 1);
    @memcpy(ifname[0..copy_len], name_buf[0..copy_len]);
    ifname[copy_len] = 0;

    return 0;
}

// ============================================================================
// System Command Execution using std.process.Child (Pure Zig)
// ============================================================================

var allocator: std.mem.Allocator = undefined;

fn run_command(argv: []const []const u8) !c_int {
    var child = std.process.Child.init(argv, allocator);
    child.stdin_behavior = .Ignore;
    child.stderr_behavior = .Ignore;

    const term = try child.spawnAndWait();
    switch (term) {
        .Exited => |code| return code,
        else => return -1,
    }
}

fn delete_route() c_int {
    return run_command(&[_][]const u8{ "route", "-q", "-n", "delete", "-inet", "10.0.0.2" }) catch return -1;
}

fn verify_route() c_int {
    return run_command(&[_][]const u8{ "route", "-n", "get", "10.0.0.2" }) catch return -1;
}

fn add_route(tun_name: [*:0]const u8) c_int {
    const tun_name_z: [*:0]const u8 = @ptrCast(tun_name);
    return run_command(&[_][]const u8{ "route", "-q", "-n", "add", "-inet", "10.0.0.2/32", "-iface", std.mem.sliceTo(tun_name_z, 0) }) catch return -1;
}

// Configure interface using ifconfig
fn configure_interface(ifname: [*:0]const u8, ip: [*:0]const u8, peer: [*:0]const u8) c_int {
    const ifname_str = std.mem.sliceTo(ifname, 0);
    const ip_str = std.mem.sliceTo(ip, 0);
    const peer_str = std.mem.sliceTo(peer, 0);

    return run_command(&[_][]const u8{ "ifconfig", ifname_str, ip_str, peer_str }) catch return -1;
}

// ============================================================================
// Socket Functions (Pure Zig using std.posix)
// ============================================================================

fn socket_create() c_int {
    const fd = std.posix.socket(std.posix.AF.INET, std.posix.SOCK.DGRAM, 0) catch return -1;
    return fd;
}

fn socket_create_sys() c_int {
    const fd = std.posix.socket(macos.PF_SYSTEM, macos.SOCK_DGRAM, macos.SYSPROTO_CONTROL) catch return -1;
    return fd;
}

fn socket_close_zig(sock: c_int) void {
    std.posix.close(sock);
}

// ============================================================================
// Connect Function (Pure Zig using std.posix.connect)
// ============================================================================

fn connect_utun(sock: c_int, ctl_id: u32) c_int {
    const addr = macos.sockaddr_ctl.init(ctl_id, 0); // sc_unit = 0 (auto-assign)
    std.posix.connect(sock, @as(*const std.posix.sockaddr, @ptrCast(&addr)), @sizeOf(macos.sockaddr_ctl)) catch return -1;
    return 0;
}

// ============================================================================
// Global Buffer
// ============================================================================

var packet_buf: [BUF_SIZE]u8 = undefined;

// ============================================================================
// IP to String (Pure Zig)
// ============================================================================

var ip2str_buf1: [16]u8 = undefined;
var ip2str_buf2: [16]u8 = undefined;
var ip2str_use_first = true;

pub fn ip2str(ip: u32) [*:0]const u8 {
    const b0 = (ip >> 24) & 0xFF;
    const b1 = (ip >> 16) & 0xFF;
    const b2 = (ip >> 8) & 0xFF;
    const b3 = ip & 0xFF;

    const buf = if (ip2str_use_first) &ip2str_buf1 else &ip2str_buf2;
    ip2str_use_first = !ip2str_use_first;

    const len = std.fmt.bufPrint(buf, "{d}.{d}.{d}.{d}", .{ b0, b1, b2, b3 }) catch unreachable;
    buf[len.len] = 0;
    return @as([*:0]const u8, @ptrCast(buf));
}

// ============================================================================
// Socket I/O (Pure Zig using C read/write)
// ============================================================================

fn set_nonblocking(fd: c_int) c_int {
    const flags = std.posix.fcntl(fd, std.posix.F.GETFL, 0) catch return -1;
    _ = std.posix.fcntl(fd, std.posix.F.SETFL, flags | 0x0004) catch return -1;
    return 0;
}

fn tun_read(fd: c_int, error_code: *c_int) isize {
    const n = read(fd, &packet_buf, BUF_SIZE);
    if (n < 0) {
        error_code.* = errno;
        return -1;
    }
    error_code.* = 0;
    return @as(isize, @intCast(n));
}

fn tun_write(fd: c_int, len: isize, error_code: *c_int) isize {
    const n = write(fd, &packet_buf, @as(usize, @intCast(len)));
    if (n < 0) {
        error_code.* = errno;
        return -1;
    }
    error_code.* = 0;
    return @as(isize, @intCast(n));
}

fn tun_close(fd: c_int) void {
    std.posix.close(fd);
}

// ============================================================================
// ioctl Functions (Pure Zig using c.ioctl)
// ============================================================================

fn ioctl_get_flags(sock: c_int, ifname: [*:0]const u8, flags: *c_int) c_int {
    var ifr: macos.ifreq = .{};
    ifr.setName(ifname);

    const req = @as(c_int, @bitCast(macos.SIOCGIFFLAGS));
    const ret = c.ioctl(sock, req, &ifr);
    if (ret < 0) return -1;

    flags.* = ifr.ifr_flags;
    return 0;
}

fn ioctl_set_flags(sock: c_int, ifname: [*:0]const u8, flags: c_int) c_int {
    var ifr: macos.ifreq = .{};
    ifr.setName(ifname);
    ifr.ifr_flags = flags;

    const req = @as(c_int, @bitCast(macos.SIOCSIFFLAGS));
    const ret = c.ioctl(sock, req, &ifr);
    return if (ret < 0) -1 else 0;
}

fn ioctl_set_ip(sock: c_int, ifname: [*:0]const u8, ip: [*:0]const u8) c_int {
    var ifr: macos.ifreq = .{};
    ifr.setName(ifname);

    // Parse IP string
    var parts: [4]u8 = undefined;
    var val: u32 = 0;
    var count: usize = 0;

    var i: usize = 0;
    while (ip[i] != 0) {
        const ch = ip[i];
        if (ch == '.') {
            parts[count] = @as(u8, @intCast(val));
            val = 0;
            count += 1;
        } else {
            val = val * 10 + (ch - '0');
        }
        i += 1;
    }
    if (count == 3) {
        parts[3] = @as(u8, @intCast(val));
        ifr.ifr_addr.sin_family = @as(u8, @intCast(std.posix.AF.INET));
        ifr.ifr_addr.sin_len = @sizeOf(macos.sockaddr_in);
        ifr.ifr_addr.sin_addr = parts;
    }

    const req = @as(c_int, @bitCast(macos.SIOCSIFADDR));
    const ret = c.ioctl(sock, req, &ifr);
    return if (ret < 0) -1 else 0;
}

fn ioctl_set_peer(sock: c_int, ifname: [*:0]const u8, peer: [*:0]const u8) c_int {
    var ifr: macos.ifreq = .{};
    ifr.setName(ifname);

    // Parse IP string
    var parts: [4]u8 = undefined;
    var val: u32 = 0;
    var count: usize = 0;

    var i: usize = 0;
    while (peer[i] != 0) {
        const ch = peer[i];
        if (ch == '.') {
            parts[count] = @as(u8, @intCast(val));
            val = 0;
            count += 1;
        } else {
            val = val * 10 + (ch - '0');
        }
        i += 1;
    }
    if (count == 3) {
        parts[3] = @as(u8, @intCast(val));
        ifr.ifr_dstaddr.sin_family = @as(u8, @intCast(std.posix.AF.INET));
        ifr.ifr_dstaddr.sin_len = @sizeOf(macos.sockaddr_in);
        ifr.ifr_dstaddr.sin_addr = parts;
    }

    const req = @as(c_int, @bitCast(macos.SIOCSIFDSTADDR));
    const ret = c.ioctl(sock, req, &ifr);
    return if (ret < 0) -1 else 0;
}

fn interface_up(ifname: [*:0]const u8) c_int {
    const sock = socket_create();
    if (sock < 0) return -1;

    var flags: c_int = 0;
    if (ioctl_get_flags(sock, ifname, &flags) < 0) {
        socket_close_zig(sock);
        return -1;
    }

    flags |= macos.IFF_UP | macos.IFF_RUNNING;

    if (ioctl_set_flags(sock, ifname, flags) < 0) {
        socket_close_zig(sock);
        return -1;
    }

    socket_close_zig(sock);
    std.debug.print("Interface up\n", .{});
    return 0;
}

// ============================================================================
// Create UTUN Socket
// ============================================================================

fn create_utun_socket(ifname: [*]u8, max_len: usize) c_int {
    const sock = socket_create_sys();
    if (sock < 0) return -1;

    var ctl_id: u32 = 0;
    if (ioctl_get_ctl_info(sock, "com.apple.net.utun_control", &ctl_id) < 0) {
        socket_close_zig(sock);
        return -1;
    }

    if (connect_utun(sock, ctl_id) < 0) {
        socket_close_zig(sock);
        return -1;
    }

    if (getsockopt_ifname(sock, ifname, max_len) < 0) {
        socket_close_zig(sock);
        return -1;
    }

    const ifname_z: [*:0]const u8 = @ptrCast(@as([*]u8, @alignCast(ifname)));
    std.debug.print("Created utun socket: {s}\n", .{ifname_z});
    return sock;
}

// ============================================================================
// Checksum (Pure Zig)
// ============================================================================

pub fn calc_sum(addr: [*]u16, len: c_int) u16 {
    var nleft = len;
    var sum: u32 = 0;
    var w = addr;

    while (nleft > 1) {
        sum += w.*;
        w += 1;
        nleft -= 2;
    }
    if (nleft == 1) {
        sum += @as(u32, w[0] & 0xFF);
    }
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    return @as(u16, @truncate(~sum));
}

// ============================================================================
// Packet Processing (Pure Zig)
// ============================================================================

pub fn process_packet(n: isize) !isize {
    const buf = &packet_buf;

    var offset: usize = 0;
    if (n >= 4 and buf[0] == 0 and buf[1] == 0) {
        offset = 4;
        std.debug.print("Skipped 4-byte utun header\n", .{});
    }

    if (n - @as(isize, @intCast(offset)) < 20) {
        std.debug.print("Packet too small\n\n", .{});
        return 0;
    }

    const vhl = buf[offset];
    const ip_hlen = (@as(usize, vhl) & 0x0F) * 4;
    const ip_len = std.mem.readInt(u16, buf[offset + 2..][0..2], .big);
    const proto = buf[offset + 9];
    const src_ip = std.mem.readInt(u32, buf[offset + 12..][0..4], .big);
    const dst_ip = std.mem.readInt(u32, buf[offset + 16..][0..4], .big);

    std.debug.print("IP: {s} -> {s}\n", .{ ip2str(src_ip), ip2str(dst_ip) });
    std.debug.print("Proto: {d} (ICMP=1)\n", .{proto});

    if (proto != 1) {
        std.debug.print("Not ICMP, skipping\n\n", .{});
        return 0;
    }

    const icmp_type = buf[offset + ip_hlen];
    std.debug.print("ICMP Type: {d} (8=echo, 0=reply)\n", .{icmp_type});

    if (icmp_type != 8) {
        std.debug.print("Not echo request, skipping\n\n", .{});
        return 0;
    }

    const icmp_id = std.mem.readInt(u16, buf[offset + ip_hlen + 4..][0..2], .big);
    const icmp_seq = std.mem.readInt(u16, buf[offset + ip_hlen + 6..][0..2], .big);
    std.debug.print("Echo Request! ID=0x{X:0>4} Seq={d}\n", .{ icmp_id, icmp_seq });

    std.mem.writeInt(u32, buf[offset + 12..][0..4], dst_ip, .big);
    std.mem.writeInt(u32, buf[offset + 16..][0..4], src_ip, .big);
    buf[offset + ip_hlen] = 0; // ICMP_ECHOREPLY

    // IP checksum
    buf[offset + 10] = 0;
    buf[offset + 11] = 0;
    var sum: u32 = 0;
    const ip_words = ip_hlen / 2;
    var i: usize = 0;
    while (i < ip_words) : (i += 1) {
        sum += std.mem.readInt(u16, buf[offset + i * 2..][0..2], .big);
    }
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    const ip_checksum = @as(u16, @truncate(~sum));
    buf[offset + 10] = @as(u8, @truncate(ip_checksum >> 8));
    buf[offset + 11] = @as(u8, @truncate(ip_checksum & 0xFF));

    // ICMP checksum
    const icmp_len = ip_len - ip_hlen;
    buf[offset + ip_hlen + 2] = 0;
    buf[offset + ip_hlen + 3] = 0;
    sum = 0;
    const icmp_words = icmp_len / 2;
    i = 0;
    while (i < icmp_words) : (i += 1) {
        sum += std.mem.readInt(u16, buf[offset + ip_hlen + i * 2..][0..2], .big);
    }
    if (icmp_len % 2 == 1) {
        sum += @as(u32, buf[offset + ip_hlen + icmp_len - 1]) << 8;
    }
    sum = (sum >> 16) + (sum & 0xFFFF);
    sum += (sum >> 16);
    const icmp_checksum = @as(u16, @truncate(~sum));
    buf[offset + ip_hlen + 2] = @as(u8, @truncate(icmp_checksum >> 8));
    buf[offset + ip_hlen + 3] = @as(u8, @truncate(icmp_checksum & 0xFF));

    std.debug.print("Reply: {s} -> {s}\n\n", .{ ip2str(dst_ip), ip2str(src_ip) });

    return n;
}

// ============================================================================
// Main
// ============================================================================

pub fn main() !void {
    allocator = std.heap.page_allocator;

    var tun_fd: c_int = undefined;
    var tun_name: [64]u8 = undefined;
    var n: isize = undefined;
    var err: c_int = undefined;

    std.debug.print("=== TUN ICMP Echo Reply Test (100% Pure Zig) ===\n\n", .{});

    _ = delete_route();

    const tun_name_ptr: [*]u8 = &tun_name;
    tun_fd = create_utun_socket(tun_name_ptr, tun_name.len);
    if (tun_fd < 0) {
        std.debug.print("Failed to create utun\n", .{});
        return error.FailedToCreateUtun;
    }

    const tun_name_z: [*:0]const u8 = @ptrCast(&tun_name);

    _ = configure_interface(tun_name_z, "10.0.0.1", "10.0.0.2");

    _ = add_route(tun_name_z);

    _ = verify_route();

    _ = set_nonblocking(tun_fd);

    std.debug.print("\nListening for ICMP...\n", .{});
    std.debug.print("(Press Ctrl+C to stop)\n\n", .{});

    while (true) {
        n = tun_read(tun_fd, &err);
        if (n < 0) {
            if (err == 35) {
                std.time.sleep(10 * std.time.ns_per_ms);
                continue;
            }
            std.debug.print("Read error: {d}\n", .{err});
            break;
        }

        std.debug.print("=== Received {d} bytes ===\n", .{n});

        const result = process_packet(n) catch {
            std.debug.print("Process error\n", .{});
            break;
        };
        if (result == 0) continue;

        n = tun_write(tun_fd, n, &err);
        if (n < 0) {
            std.debug.print("Write error: {d}\n", .{err});
        } else {
            std.debug.print("Sent {d} bytes\n\n", .{n});
        }
    }

    tun_close(tun_fd);
}
